Correctly handle loading subframes in WebArchives

sysrqb opened the pull request:
#### ba5c3db97ed13d947929aa07884cd6db17433b56
<pre>
Correctly handle loading subframes in WebArchives
<a href="https://bugs.webkit.org/show_bug.cgi?id=247292">https://bugs.webkit.org/show_bug.cgi?id=247292</a>
rdar://101776770

Reviewed by NOBODY (OOPS!).

* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba5c3db97ed13d947929aa07884cd6db17433b56

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105659 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165991 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5481 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34122 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88489 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101480 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101771 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4062 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82717 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31085 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87814 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73918 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39851 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37526 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20687 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42122 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43293 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/44012 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39951 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/PolicyChecker.cpp:
```diff
@@ -177,6 +177,14 @@ void FrameLoader::PolicyChecker::checkNavigationPolicy(ResourceRequest&& request
+#if ENABLE(WEB_ARCHIVE) && USE(CF)
+    constexpr auto webArchivePrefix = "webarchive+"_s;
+    auto url = request.url();
+    if (url.protocol().startsWith(webArchivePrefix))
```
Is this really the best place to do this? What check was actually failing before your fix?
If it is a canHandleProtocol check, I am wondering if it wouldn't be better to strip `webarchive+` in that canHandleProtocol check, so that the fix is more localized.

sysrqb replied:
Ideally we should put this in the canHandleRequest check, but that wasn't sufficient when I tested it, so I believe this is the best place. If we delay stripping it, then we may never have the opportunity to strip that prefix before a navigation decision is made. For example, if a navigation delegate is defined and it has a navigation action decision handler, then WebKit won't check if it can handle the request.

cdumez reviewed (commented):
Please write a proper commit log with an explanation.

sysrqb committed:
Correctly handle loading subframes in WebArchives
https://bugs.webkit.org/show_bug.cgi?id=247292
rdar://101776770

Reviewed by NOBODY (OOPS!).

* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):

brentfulgham reviewed (commented):
LGTM.
